wgengine/magicsock/magicsock.go

```diff
@@ -2339,6 +2336,30 @@ func (c *Conn) bindSocket(ruc *RebindingUDPConn, network string, curPortFate cur
 	return fmt.Errorf("failed to bind any ports (tried %v)", ports)
 }
 
+// getSyscallConn retrieves a raw connection out of Conn corresponding to the
+// string network, which must be "udp4" or "udp6". If it cannot get an
+// underlying UDPConn, it returns nil and errNoActiveUDP. Otherwise it returns
+// the result of SyscallConn() on the UDPConn.
+func (c *Conn) getSyscallConn(network string) (syscall.RawConn, error) {
```

Comment:
We might want to make sure that we enforce that c.pconn4 and c.pconn6 always implement syscall.Conn, and then implement syscall.Conn as needed on our custom conn types.

If we do that we could make this method `connControl(network string, fn func(fd int) error) error` and have it do along the lines of:

```
func (c *Conn) connControl(network string, fn func(fd int) error) error {
  pconn := c.pconn4
  if network == "udp6" {
    pconn = c.pconn6
  }
  sc, ok := pconn.(syscall.Conn)
  if !ok {
    return errors.New("unsupported connection type")
  }
  rc, err := sc.SyscallConn()
  if err != nil {
    return err
  }
  return rc.Control(fn)
}
```

Reply:
Thank you, I wanted to do that but didn't know how! You're now a co-author on this patch. :)

Reply:
I wasn't sure what you meant by "enforce that c.pconn4 and c.pconn6 always implement syscall.Conn". I implemented it as a cast of the underlying pconn to syscall.Conn as above in batchingUDPConn and RebindingUDPConn and returning errUnsupportedConnType if it doesn't succeed, but I could also have extended net.Packetconn to add a Syscall() method I think which would do the check at compile time? Lemme know what you had in mind.